mitodl/ocw-studio
# publish websites in root website

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/1698

#### What's this PR do?
This PR automatically creates or updates `WebsiteContent` objects of type "website" associated with the root website, denoted by `settings.ROOT_WEBSITE_NAME`. This content will be used to build an "offline index" of courses for mirror drives. There are a few components:

 - The `update_websites_in_root_website` management command, which triggers a task of the same name. The task fetches every `Website` that has been published at least once and creates a `WebsiteContent` of type "website" for it, associated with the `Website` denoted by `settings.ROOT_WEBSITE_NAME`. In order to better facilitate this, the [`django-bulk-update-or-create`](https://github.com/fopina/django-bulk-update-or-create) package was added. This allows us to bulk update or create large numbers of objects, saving significant time as opposed to just looping and running `update_or_create`.
 - The `update_website_in_root_website` task, which is meant to be set in `PREPUBLISH_ACTIONS`. This assures that when a website is published, a `WebsiteContent` object of type "website" associated with the root website is created or updated. If the root website does not have other pending content to be published, it is published as well.
 - If a site is unpublished, the `remove_website_in_root_website` task is triggered, which will remove the above described `WebsiteContent` object from the root website.

#### How should this be manually tested?
 - Spin up `ocw-studio` on this branch, ensuring you have Minio, Concourse and a custom Github org configured
 - Run `./manage.py update_websites_in_root_website`
 - Once the command has completed successfully, log in to the `ocw-studio` UI at http://localhost:8043/ publish your root website (`ocw-www` by default)
 - Once the root website is published, visit its git repository in your custom Github org and verify that you have a folder at `content/websites` on the `main`, `preview` and `live` branches that contains an entry for each published site in your database
 - Visit any course site and make an edit to the site metadata, then save your changes and publish the site live
 - Back in Github, once the course publish has finished, make sure that its Github repo has been updated as well as its entry in the root website`s Github repo
 - Unpublish the course site, either through the UI or the `unpublish_sites` management command
 - Afterwards, publish the root website again
 - Verify in the site has been unpublished and that the site's `content/websites` entry has been removed from the root website's Github repo
